proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -1012,6 +1012,268 @@ lemma refillUpdate_corres:
+               in corres_split'
+         ; (solves wpsimp)?)
+   apply (corressimp corres: refillPopHead_corres)
+  apply (subst non_overlapping_merge_refills_rewrite)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split'[OF updateRefillHd_corres])
+        apply blast
+       apply (clarsimp simp: refill_map_def)
+      apply (fastforce intro: updateRefillHd_corres
+                        simp: refill_map_def)
+     apply (wpsimp simp: update_refill_hd_def wp: update_sched_context_wp)
+     apply (clarsimp simp: vs_all_heap_simps active_sc_def is_active_sc2_def obj_at_def opt_map_def)
+    apply (wpsimp simp: updateRefillHd_def simp: objBits_simps)
+   apply (simp add: is_active_sc_rewrite[symmetric])
+  apply blast
+  done
+
+lemma head_insufficient_simp:
+  "sc_at scp s
+   \<Longrightarrow> head_insufficient scp s = Some (sc_at_pred (\<lambda>sc. r_amount (refill_hd sc) < MIN_BUDGET) scp s)"
+  unfolding head_insufficient_def
+  by (clarsimp simp: obind_def read_sched_context_def obj_at_def is_sc_obj sc_at_pred_n_def)
+
+lemma refillHdInsufficient_simp:
+  "sc_at' scp s
+   \<Longrightarrow> refillHdInsufficient scp s
+       = Some (obj_at' (\<lambda>sc :: sched_context. rAmount (refillHd sc) < minBudget) scp s)"
+  unfolding refillHdInsufficient_def
+  apply (clarsimp simp: obind_def readSchedContext_def split: option.splits)
+  apply (frule no_ofailD[OF no_ofail_sc_at'_readObject])
+  apply (clarsimp simp: obj_at'_def dest!: readObject_misc_ko_at')
+  done
+
+lemma head_insufficient_equiv:
+  "\<lbrakk>(s, s') \<in> state_relation; pspace_aligned s; pspace_distinct s; valid_objs s;
+    pred_map (\<lambda>cfg. scrc_refills cfg \<noteq> []) (sc_refill_cfgs_of s) scPtr; valid_objs' s'\<rbrakk>
+   \<Longrightarrow> head_insufficient scPtr s = refillHdInsufficient scPtr s'"
+  apply (prop_tac "sc_at scPtr s")
+   apply (fastforce dest: valid_objs_valid_sched_context_size
+                    simp: vs_all_heap_simps obj_at_kh_kheap_simps is_sc_obj_def)
+  apply (frule state_relation_pspace_relation)
+  apply (frule sc_at_cross; simp?)
+  apply (frule state_relation_sc_relation; simp?)
+  apply (subst head_insufficient_simp; simp?)
+  apply (subst refillHdInsufficient_simp; simp)
+  apply (frule refill_hd_relation2[where s'=s'])
+    apply (fastforce simp: vs_all_heap_simps opt_map_def)
+   apply (clarsimp simp: sc_ko_at_valid_objs_valid_sc' opt_map_def projectKOs obj_at'_def)
+  apply (clarsimp simp: obj_at_def sc_at_ppred_def obj_at'_def projectKOs minBudget_def
+                        MIN_BUDGET_def kernelWCETTicks_def opt_map_def vs_all_heap_simps)
+  done
+
+lemma set_object_no_fail[wp]:
+  "no_fail (obj_at (\<lambda>k. a_type obj = a_type k) sc_ptr) (set_object sc_ptr obj)"
+  apply (clarsimp simp: set_object_def)
+  apply (wpsimp wp: no_fail_obj_at get_object_wp)
+  apply (clarsimp simp: obj_at_def)
+  done
+
+lemma update_sched_context_no_fail:
+  "no_fail (sc_at sc_ptr) (update_sched_context sc_ptr f)"
```

Comment:
I think that this one could be `[wp]` as well?

Reply:
So `sc_at` wants `valid_sched_context_size` as well, but this isn't necessary here. I went with just `sc_at` for this rule because it avoided some messing around with quantifiers. I can look into it more if you think it's a good idea